src/protections/DetectFederationLag.ts

```diff
@@ -0,0 +1,744 @@
+        // Used to avoid being misled by Mjölnir catching up with old messages on first sync.
+        initialDelayGraceMS: new NumberProtectionSetting(DEFAULT_INITIAL_DELAY_GRACE_MS, 0),
+        cleanupPeriodMS: new NumberProtectionSetting(DEFAULT_CLEANUP_PERIOD_MS, 1),
+    };
+    // The instant at which the first message was received.
+    private firstMessage: Date | null = null;
+    // The latest instant at which we have started cleaning up old data.
+    private latestCleanup: Date = new Date(0);
+    private latestHistogramSettings: HistogramSettings;
+    constructor() {
+        super();
+        // Initialize and watch `this.latestHistogramSettings`.
+        this.updateLatestHistogramSettings();
+        this.settings.bucketDurationMS.on("set", () => this.updateLatestHistogramSettings());
+        this.settings.bucketNumber.on("set", () => this.updateLatestHistogramSettings());
+    }
+    dispose() {
+        this.settings.bucketDurationMS.removeAllListeners();
+        this.settings.bucketNumber.removeAllListeners();
+    }
+    public get name(): string {
+        return 'DetectFederationLag';
+    }
+    public get description(): string {
+        return `Warn moderators if either the local homeserver starts lagging by ${this.settings.localHomeserverLagEnterWarningZoneMS.value}ms or at least ${this.settings.numberOfLaggingFederatedHomeserversEnterWarningZone.value} start lagging by at least ${this.settings.federatedHomeserverLagEnterWarningZoneMS.value}ms.`;
+    }
+
+    /**
+     * @param now An argument used only by tests, to simulate events taking place at a specific date.
+     * @returns 
+     */
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any, now: Date = new Date()): Promise<any> {
+        // First, handle all cases in which we should ignore the event.
+        if (!this.firstMessage) {
+            this.firstMessage = now;
+        }
+        if (this.firstMessage.getTime() + this.settings.initialDelayGraceMS.value > now.getTime()) {
+            // We're still in the initial grace period, ignore.
+            return;
+        }
+        if (this.latestCleanup.getTime() + this.settings.cleanupPeriodMS.value > now.getTime()) {
+            // We should run some cleanup.
+            this.latestCleanup = now;
+            this.cleanup(now);
+        }
+        if (this.settings.ignoreRooms.value.has(roomId)) {
+            // Room is ignored.
+            return;
+        }
+        const sender = event['sender'] as string;
+        if (typeof sender !== "string") {
+            // Ill-formed event.
+            return;
+        }
+        if (sender === await mjolnir.client.getUserId()) {
+            // Let's not create loops.
+            return;
+        }
+        const domain = new UserID(sender).domain;
+        if (!domain) {
+            // Ill-formed event.
+            return;
+        }
+
+        const origin = event['origin_server_ts'] as number;
```

Comment:
I assume servers can falsify this timestamp. would there be use for a list of ignored spurious homeservers?

Reply:
There is a set of ignored servers. Do you want me to add a second set?